### ISSUE TITLE
ValidationError: Trying to set Table.Column of type 'number' to '10' of type 'string' at Function.fromWrongPropertyType

### ISSUE THREAD

**Issue by soroushjp** [mikro-orm/mikro-orm#324]
**Describe the bug**
With `strict: true` option provided, ORM is throwing a validation error on _read_ even though I am not setting anything.

**Stack trace**
```
ValidationError: Trying to set Stock.quantity of type 'number' to '10' of type 'string'
    at Function.fromWrongPropertyType (/path/to/my_project/node_modules/mikro-orm/dist/utils/ValidationError.js:22:16)
    at EntityValidator.validateProperty (/path/to/my_project/node_modules/mikro-orm/dist/entity/EntityValidator.js:36:43)
    at /path/to/my_project/node_modules/mikro-orm/dist/entity/EntityAssigner.js:31:49
    at Array.forEach (<anonymous>)
    at Function.assign (/path/to/my_project/node_modules/mikro-orm/dist/entity/EntityAssigner.js:16:27)
    at EntityManager.merge (/path/to/my_project/node_modules/mikro-orm/dist/EntityManager.js:219:33)
    at EntityManager.find (/path/to/my_project/node_modules/mikro-orm/dist/EntityManager.js:81:33) {
  entity: undefined,
  name: 'ValidationError'
}
```

**To Reproduce**
1. Create an entity with TS type `number`, DB type `numeric` in Postgres 12.
2. Insert some example rows into DB
3. Run `orm.em.find(models.SomeEntity, {});`

**Expected behavior**
No `ValidationError` should be triggered on read if DB schema matches TS entity (it does in my case).

**Additional context**
My DB schema was in place before using Mikro-ORM and the DDL was not generated by Mikro-ORM. Still, the column type is `numeric` and the entity field type is `number`, so there should be no issue with this column value being read in.

For now, I am avoiding this by turning off `strict: true`, which suppresses the error.

**Versions**

| Dependency | Version |
| - | - |
| node | 12.14.1 |
| typescript | 3.7.5 |
| mikro-orm | 3.1.0 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
> Create an entity with TS type number, DB type numeric in Postgres 12.

The numeric type here will be the problem I guess. PG driver is returning that as a string by default but we can easily override that.

**Comment by soroushjp** [comment]
Without knowing the details of the PG driver, one reason it could be returning it as a string is because the `numeric` type handles much greater precision (up to ~150,000 digits of precision) than JavaScript's `number`. I think your fix will work well for 95% of use cases, though it will cause issues for the edge case of very large or high precision numbers.

I can come back to help tackle this as we run into large numbers in our system (which may very possibly happen).

**Comment by soroushjp** [comment]
Also I forgot to say -- thank you for the quick fix!

**Comment by B4nan** [comment]
Yes its exactly because of that. I was thinking about making this change optional, but it should be possible to override it back to the original behaviour with the very same way (just after the ORM init happens) so I think its fine like this:

```typescript
import { types } from 'pg';

const orm = await MikroORM.init();
types.setTypeParser(1700, str => str);
```

**Comment by stephenh** [comment]
Fwiw I had a similar issue in a "generated TS clients for protobuf" project where protobuf supports longs / 64 bit numbers that are bigger than number.

At first I didn't want to support "anything not a `number`", so did something somewhat similar to @B4nan's fix for this issue, _however_, I was able to do a bounds check on the protobuf number and throw a runtime error if it was higher/lower than the JS `number` could represent.

It would be ideal to do something like this b/c right now Mikro will silently throw away precision/something that we should assume users care about (why else are they using numeric?) without telling them, and so potentially creating silent bugs.

Granted, I'm not sure if that's easily possible to do bounds checking given the PG driver just gives you a string.

For my protobuf project, eventually someone came along who really wanted/needed the "bigger than `number`" support, and they added a built-time / config flag that changed to use the `long` npm library instead of `number`s for any field marked as 64-bit in protobuf (or say `numeric` here in pg).

That enabled the "best" of both worlds: people would really wanted `number`s could still use them, but they'd get runtime errors if the client was going to actively loose/mis-represent data, and people that knew they needed the bigger-than-`number` support could opt-in to the less-idiomatic-but-it-works `Long` types.

https://www.npmjs.com/package/long

Disclaimer I've not looked into the pg `numeric` type at all, but just relating what worked well on another project for a similar-ish scenario.

**Comment by B4nan** [comment]
> At first I didn't want to support "anything not a number", so did something somewhat similar to @B4nan's fix for this issue, however, I was able to do a bounds check on the protobuf number and throw a runtime error if it was higher/lower than the JS number could represent.

Well that is kinda the way how it works now (without this fix), without strict mode enabled, string numbers will get automatically converted to number and then they are converted back to string, checking that the original string is equal. 


As said earlier, pg's numeric type is like a big float:

> up to 131072 digits before the decimal point; up to 16383 digits after the decimal point

**Comment by vowfood** [comment]
@B4nan  Just discovered something and would appreciate your guidance on good next steps.

If you re-read the thread above, we chose to coerce Postgres’ `numeric` type (which can have arbitrary precision and is returned by `pg` as a string) into a number using `parseFloat`. We spoke about the fact that we could use a custom type to avoid this coercion if we wanted to retain the arbitrary precision in our JS code. I am now attempting to do exactly that, to be able to handle floats larger than the limits of JS’ `number` type.

This is the custom type I put together for that:

```typescript
import Decimal from 'decimal.js';
import { Platform, Type } from 'mikro-orm';

export class DecimalType extends Type {

  convertToDatabaseValue(value: Decimal, _platform: Platform): string {
    return value.toFixed();
  }

  convertToJSValue(value: any, _platform: Platform): any {
    return new Decimal(value);
  }
}
```

This works great for JS -----> DB insertion.

However, for JS <---- DB retrieval, it does not work, because the previous fix causes the value argument of `convertToJSValue` to _already be coerced into a number_ prior to the call to the custom type’s `convertToJSValue`.

You can test and confirm this in Node 12 by trying to retrieve a Postgres numeric of value `9007199254740991001` — this is coerced to be `9007199254740991000` (a different number) by the time it arrives at `convertToJSValue`. 

I am going to make a change to my project’s version of MikroORM to patch this temporarily for the purposes of my project, but would love to understand how we can fix this more permanently within the official MikroORM project. Thanks!

**Comment by vowfood** [comment]
@B4nan  Just discovered something and would appreciate your guidance on good next steps.

If you re-read the thread above, we chose to coerce Postgres’ `numeric` type (which can have arbitrary precision and is returned by `pg` as a string) into a number using `parseFloat`. We spoke about the fact that we could use a custom type to avoid this coercion if we wanted to retain the arbitrary precision in our JS code. I am now attempting to do exactly that, to be able to handle floats larger than the limits of JS’ `number` type.

This is the custom type I put together for that:

```typescript
import Decimal from 'decimal.js';
import { Platform, Type } from 'mikro-orm';

export class DecimalType extends Type {

  convertToDatabaseValue(value: Decimal, _platform: Platform): string {
    return value.toFixed();
  }

  convertToJSValue(value: any, _platform: Platform): any {
    return new Decimal(value);
  }
}
```

This works great for JS -----> DB insertion.

However, for JS <---- DB retrieval, it does not work, because the previous fix causes the value argument of `convertToJSValue` to _already be coerced into a number_ prior to the call to the custom type’s `convertToJSValue`.

You can test and confirm this in Node 12 by trying to retrieve a Postgres numeric of value `9007199254740991001` — this is coerced to be `9007199254740991000` (a different number) by the time it arrives at `convertToJSValue`. 

I am going to make a change to my project’s version of MikroORM to patch this temporarily for the purposes of my project, but would love to understand how we can fix this more permanently within the official MikroORM project. Thanks!

**Comment by soroushjp** [comment]
@B4nan  Just discovered something and would appreciate your guidance on good next steps.

If you re-read the thread above, we chose to coerce Postgres’ `numeric` type (which can have arbitrary precision and is returned by `pg` as a string) into a number using `parseFloat`. We spoke about the fact that we could use a custom type to avoid this coercion if we wanted to retain the arbitrary precision in our JS code. I am now attempting to do exactly that, to be able to handle floats larger than the limits of JS’ `number` type.

This is the custom type I put together for that:

```typescript
import Decimal from 'decimal.js';
import { Platform, Type } from 'mikro-orm';

export class DecimalType extends Type {

  convertToDatabaseValue(value: Decimal, _platform: Platform): string {
    return value.toFixed();
  }

  convertToJSValue(value: any, _platform: Platform): any {
    return new Decimal(value);
  }
}
```

This works great for JS -----> DB insertion.

However, for JS <---- DB retrieval, it does not work, because the previous fix causes the value argument of `convertToJSValue` to _already be coerced into a number_ prior to the call to the custom type’s `convertToJSValue`.

You can test and confirm this in Node 12 by trying to retrieve a Postgres numeric of value `9007199254740991001` — this is coerced to be `9007199254740991000` (a different number) by the time it arrives at `convertToJSValue`. 

I am going to make a change to my project’s version of MikroORM to patch this temporarily for the purposes of my project, but would love to understand how we can fix this more permanently within the official MikroORM project. Thanks!

**Comment by B4nan** [comment]
I thought you can override this easily, right after you initialize the ORM, you can do:

```
import { types } from 'pg';
types.setTypeParser(1700, str => str);
```

**Comment by soroushjp** [comment]
^ Can confirm this works when called after `MikroORM.init`, thanks @B4nan. Avoids needing to modify my project's copy of MikroORM, which is great.

Do you want library users to be doing this sort of thing, of reaching into the underlying DB drivers? I know you have `driverOptions` available, but this is somewhat different to that.

**Comment by B4nan** [comment]
Yeah converting the value to number was a bad idea, let's drop that conversion in v4 and let users do what they want with the string (probably via predefined custom type).

**Comment by soroushjp** [comment]
I think an alternative would be to expose this as a config in `MikroORM.init`, so that users can turn it off via an explicit configuration rather than by reaching into the driver. Could it somehow be included in `driverOptions` or somewhere similar?

**Comment by soroushjp** [comment]
Actually on second thought, I agree with you, let's keep it an explicit decision by the user. If they use an `int4`, pg returns this as a `number` as expected. It's only larger Postgres number types such as `int8` and `numeric` that are returned as `string`, and it's good for the caller to explicitly consider the implications of this for their software.

Any reason we want to keep this type conversion in v3 versus removing it? It wasn't present in earlier versions of v3 so it's not exactly a breaking change.

**Comment by B4nan** [comment]
Well in fact the change I made here before was a breaking change too, now I will be basically reverting it :D The change was part of 3.1.1, I feel like reverting it in late 3.6 version is not that good idea (although small BC, still a BC, and doing it at the end of life of v3 feels wrong). 

Generally I think I am starting to take more things into the BC category gradually as the user base is growing. And v4 will come quite soon, so not that big deal hopefully, especially because there is that workaround mentioned above.

**Comment by soroushjp** [comment]
Sounds good @B4nan, that makes sense. I'm not personally bothered at all since I have the workaround. Thanks for thoughts here and hopefully something that can be resolved properly in v4.

**Comment by B4nan** [comment]
Closing as fixed in `dev` branch, subscribe to #527 for updates.
